modules/foreman_os_default_template.py

```diff
@@ -0,0 +1,204 @@
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    operatingsystem: "CoolOS"
+    template_kind: "finish"
+    provisioning_template: "CoolOS finish"
+    state: present
+
+- name: "Delete an Association"
+  foreman_operating_system:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    operatingsystem: "CoolOS"
+    template_kind: "finish"
+    state: absent
+'''
+
+RETURN = ''' # '''
+
+try:
+    from ansible.module_utils.ansible_nailgun_cement import (
+        create_server,
+        ping_server,
+        find_entities_by_name,
+        find_os_default_template,
+        naildown_entity_state,
+    )
+
+    from nailgun.entities import (
+        OperatingSystem,
+        OSDefaultTemplate,
+        ProvisioningTemplate,
+        TemplateKind,
+    )
+
+    HAS_NAILGUN_PACKAGE = True
+except:
+    HAS_NAILGUN_PACKAGE = False
+    raise
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.foreman_helper import handle_no_nailgun
+
+
+def sanitize_os_default_template_dict(entity_dict):
+    # This is the only true source for names (and conversions thereof)
+    name_map = {
+        'operatingsystem': 'operatingsystem',
+        'template_kind': 'template_kind',
+        'provisioning_template': 'provisioning_template',
+    }
+    result = {}
+    for key, value in name_map.iteritems():
+        if key in entity_dict:
+            result[value] = entity_dict[key]
+    return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(type='bool', default=True),
+            operatingsystem=dict(required=True),
+            template_kind=dict(required=True),
+            provisioning_template=dict(required=False),
+            state=dict(required=True, choices=['present', 'present_with_defaults', 'absent']),
+        ),
+        required_if=(
+            ['state', 'present', ['provisioning_template']],
+            ['state', 'present_with_defaults', ['provisioning_template']],
+        ),
+        supports_check_mode=True,
+    )
+
+    handle_no_nailgun(module, HAS_NAILGUN_PACKAGE)
+
+    entity_dict = dict(
+        [(k, v) for (k, v) in module.params.iteritems() if v is not None])
+
+    server_url = entity_dict.pop('server_url')
+    username = entity_dict.pop('username')
+    password = entity_dict.pop('password')
+    verify_ssl = entity_dict.pop('verify_ssl')
+    state = entity_dict.pop('state')
+
+    try:
+        create_server(server_url, (username, password), verify_ssl)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+    ping_server(module)
+    entity_dict['operatingsystem'] = find_entities_by_name(
+        OperatingSystem, [entity_dict['operatingsystem']], module)[0]
```

Comment:
I don't think `OperatingSystem` names are unique. We should really be doing a title search using '~':

```python
OperatingSystem.search(set(), {'search': 'title~"{}"'.format(entity_dict['operatingsystem'])})
```

This allows a distinction between, say, Debian 9.1 and Debian 9.2.